File: src/DesignPatterns/DIY/Mediator/Mediator.java
package DesignPatterns.DIY.Mediator;

public class Mediator {
    public static void main(String[] args) {
        Payroll payroll = new Payroll();
        HRDepartment hr = new HRDepartment(payroll);
        Bank bank = new Bank();

        payroll.registerHR(hr);
        payroll.registerBank(bank);
        
        hr.sendSalary("Danah", 15000);
        
    }
}

class HRDepartment{
    private Payroll payroll;

    public HRDepartment(Payroll payroll) {
        this.payroll = payroll;
    }
    public void sendSalary(String employee, double amount){
        System.out.println("HR: Sending salary info for: " + employee);
        payroll.processSalary(employee,amount);
    }

    public Payroll getPayroll() {
        return payroll;
    }

    public void setPayroll(Payroll payroll) {
        this.payroll = payroll;
    }
    
}
class Payroll{
    private HRDepartment hrDept;
    private Bank bank;

    public HRDepartment getHrDept() {
        return hrDept;
    }
    public void setHrDept(HRDepartment hrDept) {
        this.hrDept = hrDept;
    }
    public Bank getBank() {
        return bank;
    }
    public void setBank(Bank bank) {
        this.bank = bank;
    }
    public void registerHR(HRDepartment hrDept){
        this.hrDept = hrDept;
    }
    public void registerBank(Bank bank){
        this.bank = bank;
    }
    public void processSalary(String employee, double amount){
        System.out.println("Payroll: Processing salary for: " + employee);
        if(bank != null){
            bank.transferSalary(employee, amount);
        }
    }

}
class Bank{
    public void transferSalary(String employee, double amount){
        System.out.println("Bank: Transferring Php " + amount + " to " + employee + "`s account.");
    }
}

File: bin/DIY/Mediator/Mediator.ts
// Mediator.ts

class HRDepartment {
    private payroll: Payroll;

    constructor(payroll: Payroll) {
        this.payroll = payroll;
    }

    sendSalary(employee: string, amount: number): void {
        console.log("HR: Sending salary info for: " + employee);
        this.payroll.processSalary(employee, amount);
    }

    getPayroll(): Payroll {
        return this.payroll;
    }

    setPayroll(payroll: Payroll): void {
        this.payroll = payroll;
    }
}

class Payroll {
    private hrDept: HRDepartment | null = null;
    private bank: Bank | null = null;

    getHrDept(): HRDepartment | null {
        return this.hrDept;
    }

    setHrDept(hrDept: HRDepartment): void {
        this.hrDept = hrDept;
    }

    getBank(): Bank | null {
        return this.bank;
    }

    setBank(bank: Bank): void {
        this.bank = bank;
    }

    registerHR(hrDept: HRDepartment): void {
        this.hrDept = hrDept;
    }

    registerBank(bank: Bank): void {
        this.bank = bank;
    }

    processSalary(employee: string, amount: number): void {
        console.log("Payroll: Processing salary for: " + employee);
        if (this.bank !== null) {
            this.bank.transferSalary(employee, amount);
        }
    }
}

class Bank {
    transferSalary(employee: string, amount: number): void {
        console.log("Bank: Transferring Php " + amount + " to " + employee + "`s account.");
    }
}

// main
(function main() {
    const payroll = new Payroll();
    const hr = new HRDepartment(payroll);
    const bank = new Bank();

    payroll.registerHR(hr);
    payroll.registerBank(bank);

    hr.sendSalary("Danah", 15000);
})();